### ISSUE TITLE
Can snowalert send attachments from s3 in email.

### ISSUE THREAD

**Issue by syedirfanahmedansari** [snowflakedb/SnowAlert#541]
we have requirement where we want error files/csv files to be send as an attachment.

Can it be done using SES/SNS or simply SMTP handlers.

Please respond ASAP.

Thanks!
Irfan.

**Comment by sfc-gh-afedorov** [comment]
Could you clarify what you mean by "from S3"?

**Comment by syedirfanahmedansari** [comment]
Can snowalert access AWS S3 to send attachments.

**Comment by sfc-gh-afedorov** [comment]
Snowflake can access S3 via PIPE's and External Functions which can access S3 via boto3, and SnowAlert runs on Snowflake, so yes, SnowAlert can access S3 to send attachments.

**Comment by sfc-gh-afedorov** [comment]
Reviewing their code ([ses.py](https://github.com/snowflakedb/SnowAlert/blob/master/src/runners/handlers/ses.py) [smtp.py](https://github.com/snowflakedb/SnowAlert/blob/master/src/runners/handlers/smtp.py), I don't think any of the current handlers will reach out to S3 directly. It's probably best that way -- their role is to send emails, not read data from S3.

**Comment by syedirfanahmedansari** [comment]
i don't see any part in ses.py/smtp.py where they take attachments.

so this means the current handlers don't support file attachment from any where.

**Comment by sfc-gh-afedorov** [comment]
Yes, seems like you'd need to add `attachment` to the params and —

~~~python
if attachment:
  att = MIMEApplication(attachment['content'])
  att.add_header('Content-Disposition', 'attachment', filename=attachment['filename'])
  message.attach(att)
~~~

to `smtp.py` for that to work.

We're moving over to using External Functions for handlers, tho, and will have news on that front soon.

Can update the ticket when that work is ready for public release.

**Comment by syedirfanahmedansari** [comment]
at our end snowalert runs on  AWS ECR. if the attachment is to be added then that file needs to present in running machine. for attachment to work the file needs to present locally.
My concern is how will it attach from outside.
or how do i send the file to the docker instance.

**Comment by syedirfanahmedansari** [comment]
Can you send me a sample alert view for using SES handler

**Comment by sfc-gh-afedorov** [comment]
we don't have one for SES on hand (it's a third party contribution that my team doesn't use), but SMTP is documented [here](https://docs.snowalert.com/alerts/handlers/smtp). all the handlers are the same interface, tho -- the "type" key specifies which Python module is loaded, and the rest of the keys are passed in as parameters to the "handle" method in that module.

**Comment by syedirfanahmedansari** [comment]
Can you please share a sample select for SMTP handler

**Comment by sfc-gh-afedorov** [comment]
https://docs.snowalert.com/alerts/handlers/smtp

**Comment by syedirfanahmedansari** [comment]
Hi,

Can you please provide Steps to configure Snowalert on AWS using Cloud Front stack.

Thanks!
Irfan.

**Comment by sfc-gh-afedorov** [comment]
We have not deployed on cfn ourselves, but someone who has provided [this code](https://github.com/snowflakedb/SnowAlert/blob/master/infra/cfn/snowalert.yaml) and [these instructions](https://github.com/snowflakedb/SnowAlert/blob/master/infra/cfn/CLOUDFORMATION.md). Does that help?

**Comment by syedirfanahmedansari** [comment]
Hi,

it seems we are getting duplicate alerts.

the way we are running is we have a view in snowflake to which the final Alert query is pointing and we have scheduled snow alert to run for every 5 minutes.

if any entry comes in two executions then the alerts are being published twice. how can we suppress the second alert.

**Comment by sfc-gh-afedorov** [comment]
Alerts are deduplicated by `object` and `description`, so if those are identical and the alerts are within the look-back period, you should be good. Could you check whether e.g. you are including the current time in the description, perhap?
